lib/github.com/diku-dk/containers/hash.fut

```diff
@@ -0,0 +1,448 @@
+
+  #[inline]
+  def (&) (a: t) (b: t) : t =
+    { high = high a & high b
+    , mid = mid a & mid b
+    , low = low a & low b
+    }
+
+  -- | Shift by 89
+  #[inline]
+  def shift_prime (a: t) =
+    { high = 0u64
+    , mid = high a >> 25
+    , low = (mid a u64.>> 25) u64.| (high a u64.<< 39)
+    }
+
+  #[inline]
+  def (>=) (a: t) (b: t) : bool =
+    (high a) > (high b)
+    || ((high a) == (high b) && (mid a) > (mid b))
+    || ((high a) == (high b) && (mid a) == (mid b) && (low a) >= (low b))
+
+  #[inline]
+  def (==) (a: t) (b: t) : bool =
+    (high a) == (high b) && (mid a) == (mid b) && (low a) == (low b)
+
+  def n : i64 = 3
+
+  #[inline]
+  def from_u64 (as: [n]u64) : t =
+    {high = as[0], mid = as[1], low = as[2]}
+
+  #[inline]
+  def from_u (a: u) : t =
+    {high = 0u64, mid = 0u64, low = a}
+
+  #[inline]
+  def mul_small (a: t) (b: u) : t =
+    let lo = (low a) * b
+    let mi' = (mid a) * b
+    let mi = mi' + u64.mul_hi (low a) b
+    let carry = u64.bool (mi < mi')
+    let hi = (high a) * b + u64.mul_hi (mid a) b + carry
+    in {high = hi, mid = mi, low = lo}
+
+  #[inline]
+  def (*) (a: t) (b: t) : t =
+    let lo = (low a) * (low b)
+    let mi'' = (mid a) * (low b)
+    let mi' = mi'' + (mid b) * (low a)
+    let mi = mi' + u64.mul_hi (low a) (low b)
+    let carry = u64.bool (mi' < mi'') + u64.bool (mi < mi')
+    let hi =
+      (high a) * (low b)
+      + (high b) * (low a)
+      + (mid a) * (mid b)
+      + u64.mul_hi (mid a) (low b)
+      + u64.mul_hi (mid b) (low a)
+      + carry
+    in {high = hi, mid = mi, low = lo}
+
+  #[inline]
+  def (+) (a: t) (b: t) =
+    let lo = (low a) + (low b)
+    let mi' = (mid a) + (mid b)
+    let mi = mi' + u64.bool (lo < low a)
+    let hi = (high a) + (high b) + u64.bool (mi' < mid a) + u64.bool (mi < mi')
+    in { high = hi
+       , mid = mi
+       , low = lo
+       }
+
+  #[inline]
+  def (-) (a: t) (b: t) =
+    let lo = (low a) - (low b)
+    let mi' = (mid a) - (mid b)
+    let mi = mi' - u64.bool (lo > low a)
+    let hi = (high a) - (high b) - u64.bool (mi' > mid a) - u64.bool (mi > mi')
+    in { high = hi
+       , mid = mi
+       , low = lo
+       }
+
+  #[inline]
+  def to_u (n: t) : u =
+    n.low
+
+  #[inline]
+  def to_u64 (a: t) : [n]u64 =
+    sized n [a.high, a.mid, a.low]
+}
+
+type u192 = u192.t
+
+-- | n-dimensional linear congruential generator. You can give it a
+-- list of non-zero integers to generate an n-array of n random
+-- numbers that will go in a full cycle such that every n-array of
+-- will be hit. The number size is limited by the uint implementation
+-- so the numbers will be betwen zero and less than the mersenne
+-- prime. The purpose of the random number generator is mainly meant
+-- for generating numbers for a universal hash function has constants.
+module mk_ndimlcg
+  (U: uint)
+  (P: {
+    -- | The type of random numbers to use.
+    type t
+
+    -- | Number of random numbers to generate.
+    val n : i64
```

Comment:
Isn't this more "the number of dimensions"? You can always call `rand` multiple times to get more numbers.